### DISCUSSION TITLE
Unique_hash_map can not be specialized by Face_handle?

### DISCUSSION THREAD

**Discussion by xuanjueheshang** [CGAL/cgal#5961] [Q&A] [answered]
I got a problem. 

I made a voronoi diagram with the voronoi library. Then I wanted to bind some data to "some special faces". A Face_handle can be created for me when I use the "locate" methed, Then, what should i do to bind data to it? Or is there a method to convert Face_handle to Face_iterator?

Unique_hash_map can not be specialized by Face_handle. Face_iterator works well. ==! Or there is a special skill?

Sorry for my poor English.

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): Win 10
* Compiler: MSVC
* Release or debug mode: Release/Debug
* CGAL version:5.3
* Boost version:1.76

**Comment by lrineau** [comment]
1. You can modify the "face" type of the diagram to add new data members to it.
2. I am surprised by your statement "`Unique_hash_map` can not be specialized by `Face_handle`". You can please share you code and show us what does not work at you wish?

**Reply by xuanjueheshang** [reply]
thanks.
here is the code.
[main.zip](https://github.com/CGAL/cgal/files/7148601/main.zip)

**Comment by lrineau** [comment, marked as answer]
I get it.

The compilation error is:
```
.../Hash_map/include/CGAL/Handle_hash_function.h:37:11: error: no type named ‘value_type’ in ‘struct std::iterator_traits<CGAL::VoronoiDiagram_2::Internal::Handle_adaptor<CGAL::VoronoiDiagram_2::Internal::Face<CGAL::Voronoi_diagram_2<CGAL::Delaunay_triangulation_2<CGAL::Epick>, CGAL::Delaunay_triangulation_adaptation_traits_2<CGAL::Delaunay_triangulation_2<CGAL::Epick> >, CGAL::Delaunay_triangulation_caching_degeneracy_removal_policy_2<CGAL::Delaunay_triangulation_2<CGAL::Epick> > > > > >’
   37 |           sizeof( typename std::iterator_traits<H>::value_type);
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The issue is that `CGAL::VoronoiDiagram_2::Internal::Handle_adaptor<T>` is not a correct model of our [`Handle`](https://doc.cgal.org/5.3/Circulator/classHandle.html) concept. Here is a patch to fix that:
```patch
diff --git a/Voronoi_diagram_2/include/CGAL/Voronoi_diagram_2/Handle_adaptor.h b/Voronoi_diagram_2/include/CGAL/Voronoi_diagram_2/Handle_adaptor.h
index 91b5be76f09..8ed30759595 100644
--- a/Voronoi_diagram_2/include/CGAL/Voronoi_diagram_2/Handle_adaptor.h
+++ b/Voronoi_diagram_2/include/CGAL/Voronoi_diagram_2/Handle_adaptor.h
@@ -33,6 +33,8 @@ class Handle_adaptor
   typedef T&     reference;
   typedef const T*  const_pointer;
   typedef const T&  const_reference;
+  typedef void iterator_category;
+  typedef std::ptrdiff_t difference_type;
 
  public:
   Handle_adaptor() : t() {}
```

Submitted as pull-request https://github.com/CGAL/cgal/pull/5981.

**Reply by xuanjueheshang** [reply]
Thanks
